### PR TITLE
TypeScript support for client/payment-methods

### PR DESCRIPTION
**Pull request by elazzabi**
This PR has no functional changes. Only migrating part of the code to TypeScript

#### Changes proposed in this Pull Request

No real changes, only migration to TypeScript

-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. **(not needed)**
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)  **(not needed)**

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) : _Add link here / 'QA Testing Not Applicable'_

### CODE REVIEW

File: client/components/confirmation-modal/index.tsx [hunk 1]
```diff
@@ -11,9 +11,26 @@ import { HorizontalRule } from '@wordpress/primitives';
  */
 import './styles.scss';
 
-const ConfirmationModal = ( { children, actions, className, ...props } ) => (
+interface ConfirmationModalInput {
```
Comment (kalessil) [inline client/components/confirmation-modal/index.tsx:14]: I'd suggest using following (incl. naming conventions):

```
interface ConfirmationModalProps extends Modal.Props {
	actions: JSX.Element;
}
```

File: client/components/payment-confirm-illustration/index.tsx [hunk 2]
```diff
@@ -10,7 +11,14 @@ import classNames from 'classnames';
  */
 import './styles.scss';
 
-const PaymentConfirmIllustration = ( { hasBorder, Icon = () => null } ) => {
+const PaymentConfirmIllustration = ( {
+	hasBorder,
+	Icon = () => <React.Fragment />,
```
Comment (kalessil) [inline client/components/payment-confirm-illustration/index.tsx:16]: can we rename it to `icon` to comply with the naming convention?

File: client/components/payment-delete-illustration/index.tsx [hunk 3]
```diff
@@ -10,7 +11,14 @@ import classNames from 'classnames';
...
 import './styles.scss';
 
-const PaymentDeleteIllustration = ( { hasBorder, Icon = () => null } ) => {
+const PaymentDeleteIllustration = ( {
+	hasBorder,
+	Icon = () => <React.Fragment />,
+}: {
+	hasBorder: boolean;
+	// eslint-disable-next-line @typescript-eslint/naming-convention
+	Icon: ( { className }: { className: string } ) => JSX.Element;
```
Comment (kalessil) [inline client/components/payment-delete-illustration/index.tsx:20]: Similar here: `icon`

File: client/payment-methods-map.ts [hunk 4]
```diff
@@ -16,7 +16,21 @@ import P24Icon from './gateway-icons/p24';
 import IdealIcon from './gateway-icons/ideal';
 import BankDebitIcon from './gateway-icons/bank-debit';
 
-export default {
+export interface PaymentMethodInformationType {
```
Comment (kalessil) [inline client/payment-methods-map.ts:19]: Would `PaymentMethodMapEntry` be more speaking?

File: client/payment-methods-map.ts [hunk 5]
```diff
@@ -16,7 +16,21 @@ import P24Icon from './gateway-icons/p24';
...
+	Icon: () => JSX.Element;
+	currencies: string[];
+	stripe_key: string;
+}
+
+interface PaymentMethodInformationObjectType {
+	[ key: string ]: PaymentMethodInformationType;
+}
+
+const PaymentMethodInformationObject: PaymentMethodInformationObjectType = {
```
Comment (kalessil) [inline client/payment-methods-map.ts:33]: `const PaymentMethodInformationObject: Record< string, PaymentMethodMapEntry > = {`

File: client/payment-methods/activation-modal.tsx [hunk 6]
```diff
@@ -16,24 +16,33 @@ import PaymentMethodsMap from '../payment-methods-map';
...
+	onConfirmClose: () => void;
+}
+
 const ConfirmPaymentMethodActivationModal = ( {
 	paymentMethod,
 	requirements,
 	onClose,
 	onConfirmClose,
-} ) => {
+}: ConfirmaPaymentMethodActivationType ): JSX.Element => {
```
Comment (kalessil) [inline client/payment-methods/activation-modal.tsx:31]: Can we inline `ConfirmaPaymentMethodActivationType` here?

File: client/payment-methods/activation-modal.tsx [hunk 7]
```diff
@@ -16,24 +16,33 @@ import PaymentMethodsMap from '../payment-methods-map';
...
 	const requirementsToDisplay = requirements.filter( ( requirement ) => {
 		return RequirementsMap.hasOwnProperty( requirement );
 	} );
 
+	const paymentMethodInformation = PaymentMethodsMap[ paymentMethod ];
```
Comment (kalessil) [inline client/payment-methods/activation-modal.tsx:36]: Nice!

File: client/payment-methods/delete-modal.tsx [hunk 8]
```diff
@@ -12,13 +13,22 @@ import interpolateComponents from 'interpolate-components';
...
+}
+
 const ConfirmPaymentMethodDeleteModal = ( {
 	id,
 	label,
 	Icon,
 	onConfirm,
 	onCancel,
-} ) => {
+}: ConfirmPaymentMethodDeleteModalInput ): JSX.Element => {
```
Comment (kalessil) [inline client/payment-methods/delete-modal.tsx:31]: Also: inline.

File: client/requirements-map.ts [hunk 9]
```diff
@@ -3,11 +3,17 @@
  */
 import { __ } from '@wordpress/i18n';
 
-export default {
+interface RequirementsInfoObjectType {
+	[ key: string ]: string;
+}
+
+const RequirementsInfoObject: RequirementsInfoObjectType = {
```
Comment (kalessil) [inline client/requirements-map.ts:10]: `Record<string, string>`

**Review by kalessil — CHANGES_REQUESTED** [review]
Great progress on the first TS PR! I left a few comments to align on the conventions we are following.

File: client/requirements-map.ts (hunk 9 above)
Comment (elazzabi) [reply client/requirements-map.ts]: TIL! Thank you. Fixed

File: client/payment-methods/delete-modal.tsx (hunk 8 above)
Comment (elazzabi) [reply client/payment-methods/delete-modal.tsx]: Done. Is there a convention on when we should use `inline` vs a separate interface?

File: client/components/payment-confirm-illustration/index.tsx (hunk 2 above)
Comment (elazzabi) [reply client/components/payment-confirm-illustration/index.tsx:16]: I don't think it's possible given it's used as a JSX property afterward `<Icon />`, so it should start with a capital letter.

One thing we can do is make it lowercase, but once we want to render it, assign it to another variable with a capital letter. Something like `const Icon = icon;`.

It looks like a weird fix, but I agree that `props` should be lowercase.

File: client/components/confirmation-modal/index.tsx (hunk 1 above)
Comment (elazzabi) [reply client/components/confirmation-modal/index.tsx:14]: I've been looking at how I can achieve this yesterday, didn't know about the `extends Modal.props` part, thank you!

File: client/payment-methods/delete-modal.tsx (hunk 8 above)
Comment (kalessil) [reply client/payment-methods/delete-modal.tsx]: Reusable interfaces are typically separate interfaces, others tend to be inline definitions.

File: changelog/fix-1152-wcpay-checkout-payment-methods-fields-visibility [hunk 10]
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: fix
+
+Improve visibility of checkout fields for WCPay payment options on a darker theme
```
Comment (kalessil) [inline changelog/fix-1152-wcpay-checkout-payment-methods-fields-visibility:4]: We can skip the changelog in this PR, as we'll be merging our work in a dedicated branch, where the single changelog entry will be introduced.

File: client/checkout/blocks/style.scss [hunk 11]
```diff
@@ -2,6 +2,10 @@
 	margin-bottom: 2rem;
 }
 
+.wcpay-card-mounted {
+	background-color: #fff;
+}
+
```
Comment (kalessil) [inline client/checkout/blocks/style.scss:8]: Doen;t look like a TS thingy =) What's the background for introducing this change?

File: client/components/payment-confirm-illustration/index.tsx (hunk 2 above)
Comment (kalessil) [reply client/components/payment-confirm-illustration/index.tsx:16]: Okay, can you create a discussion thread in the team's private channel so we could align on if we should touch this, please?

File: client/globals.d.ts [hunk 12]
```diff
@@ -20,5 +20,6 @@ declare const wcpaySettings: {
 		country: string;
 		availableStates: Array< Record< string, string > >;
 	};
+	accountEmail: string | null;
```
Comment (kalessil) [inline client/globals.d.ts:23]: I looked into the backend and it's always a string. Falls back to empty string.

File: includes/class-wc-payments-account.php [hunk 13]
```diff
@@ -45,6 +45,7 @@ public function __construct( WC_Payments_API_Client $payments_api_client ) {
 
 		add_action( 'admin_init', [ $this, 'maybe_handle_onboarding' ] );
 		add_action( 'admin_init', [ $this, 'maybe_redirect_to_onboarding' ], 11 ); // Run this after the WC setup wizard and onboarding redirection logic.
+		add_action( 'admin_init', [ $this, 'maybe_redirect_to_wcpay_connect' ], 12 ); // Run this after the redirect to onboarding logic.
```
Comment (kalessil) [inline includes/class-wc-payments-account.php:48]: This also looks unrelated to TS migration. Perhaps the branch shall be rebased?

**Review by kalessil — CHANGES_REQUESTED** [review]
Thanks for addressing the previous review notes!
I added new more comments, but it looks like we should rebase the branch (note that the target PR branch is changed).

File: changelog/fix-1152-wcpay-checkout-payment-methods-fields-visibility (hunk 10 above)
Comment (elazzabi) [reply changelog/fix-1152-wcpay-checkout-payment-methods-fields-visibility:4]: I'm not sure from where those changes appeared 🤔 I think it's because I rebased on `develop` instead of `dev/pulsar-hackweek-ts`. Looking into it

Comment (elazzabi) [reply changelog/fix-1152-wcpay-checkout-payment-methods-fields-visibility]: Indeed. I reverted the changes and it looks much better now. Thank you for changing the base branch

File: client/components/payment-confirm-illustration/index.tsx [hunk 14]
```diff
@@ -10,7 +11,11 @@ import classNames from 'classnames';
  */
 import './styles.scss';
 
-const PaymentConfirmIllustration = ( { hasBorder, Icon = () => null } ) => {
+const PaymentConfirmIllustration: React.FunctionComponent< {
+	hasBorder: boolean;
+	// eslint-disable-next-line @typescript-eslint/naming-convention
```
Comment (kalessil) [inline client/components/payment-confirm-illustration/index.tsx:16]: I guess we can remove this suppression comment now.

File: client/components/payment-delete-illustration/index.tsx [hunk 15]
```diff
@@ -10,7 +11,11 @@ import classNames from 'classnames';
  */
 import './styles.scss';
 
-const PaymentDeleteIllustration = ( { hasBorder, Icon = () => null } ) => {
+const PaymentDeleteIllustration: React.FunctionComponent< {
+	hasBorder: boolean;
+	// eslint-disable-next-line @typescript-eslint/naming-convention
```
Comment (kalessil) [inline client/components/payment-delete-illustration/index.tsx:16]: I guess we can remove this suppression comment now.

File: client/payment-methods-map.ts [hunk 16]
```diff
@@ -16,7 +16,20 @@ import P24Icon from './gateway-icons/p24';
 import IdealIcon from './gateway-icons/ideal';
 import BankDebitIcon from './gateway-icons/bank-debit';
 
-export default {
+export interface PaymentMethodMapEntry {
+	id: string;
+	label: string;
+	description: string;
+	// eslint-disable-next-line @typescript-eslint/naming-convention
+	Icon: () => JSX.Element;
```
Comment (kalessil) [inline client/payment-methods-map.ts:24]: Not sure, if this is a leftover.

File: client/payment-methods/delete-modal.tsx [hunk 17]
```diff
@@ -18,7 +19,14 @@ const ConfirmPaymentMethodDeleteModal = ( {
 	Icon,
 	onConfirm,
 	onCancel,
-} ) => {
+}: {
+	id: string;
+	label: string;
+	// eslint-disable-next-line @typescript-eslint/naming-convention
+	Icon: () => JSX.Element;
```
Comment (kalessil) [inline client/payment-methods/delete-modal.tsx:26]: Not sure if this is a leftover.

**Review by kalessil — CHANGES_REQUESTED** [review]
Thanks for addressing previous feedback! Adding minor comments to address before we can go ahead.

File: client/payment-methods-map.ts (hunk 16 above)
Comment (elazzabi) [reply client/payment-methods-map.ts]: Good catch. I removed this, but it was a little tricky. It's being used in multiple non-TS files, so I made a global search and changed it everywhere I can find it.

**Comment by elazzabi** [comment]
Thanks for the feedback @kalessil ! I made a few changes to fix the issues you mentioned + cleaned it a little bit.

File: client/components/payment-delete-illustration/index.tsx [hunk 18]
```diff
@@ -10,7 +11,10 @@ import classNames from 'classnames';
  */
 import './styles.scss';
 
-const PaymentDeleteIllustration = ( { hasBorder, Icon = () => null } ) => {
+const PaymentDeleteIllustration: React.FunctionComponent< {
```
Comment (kalessil) [inline client/components/payment-delete-illustration/index.tsx:14]: It would be great to introduce a test for this component (a snapshot will work as we have nothing yet).

File: client/components/payment-confirm-illustration/index.tsx [hunk 19]
```diff
@@ -10,7 +11,10 @@ import classNames from 'classnames';
  */
 import './styles.scss';
 
-const PaymentConfirmIllustration = ( { hasBorder, Icon = () => null } ) => {
+const PaymentConfirmIllustration: React.FunctionComponent< {
```
Comment (kalessil) [inline client/components/payment-confirm-illustration/index.tsx:14]: It would be great to introduce a test for this component (a snapshot will work as we have nothing yet).

**Review by kalessil — APPROVED** [review]
LGTM! It'd be great though to add some basic tests for some components (see my comments).

File: client/components/payment-delete-illustration/index.tsx (hunk 18 above)
Comment (elazzabi) [reply client/components/payment-delete-illustration/index.tsx:14]: Good idea! I added tests in https://github.com/Automattic/woocommerce-payments/pull/3910/commits/473fc2521ceedd21ac3f9a731e157f578b846e93

File: client/components/payment-confirm-illustration/index.tsx (hunk 19 above)
Comment (elazzabi) [reply client/components/payment-confirm-illustration/index.tsx:14]: Added tests in https://github.com/Automattic/woocommerce-payments/pull/3910/commits/eb2f8fb875423f9788085111638972effb3a6f41
